src/java_bytecode/java_qualifiers.cpp

```diff
@@ -0,0 +1,154 @@
+// Copyright 2018 DiffBlue Limited. All Rights Reserved.
+
+/// \file
+/// Java-specific type qualifiers
+
+#include "java_qualifiers.h"
+#include <sstream>
+#include <util/make_unique.h>
+#include "expr2java.h"
+
+java_qualifierst::java_qualifierst(const namespacet &ns) : ns(ns)
+{
+}
+
+java_qualifierst &java_qualifierst::operator=(const java_qualifierst &other)
+{
+  INVARIANT(
+    &other.ns == &ns,
+    "Can only assign from a java_qualifierst using the same namespace");
+  annotations = other.annotations;
+  c_qualifierst::operator=(other);
+  return *this;
+}
+
+java_qualifierst &java_qualifierst::operator=(java_qualifierst &&other)
+{
+  INVARIANT(
+    &other.ns == &ns,
+    "Can only assign from a java_qualifierst using the same namespace");
+  annotations = std::move(other.annotations);
+  c_qualifierst::operator=(other);
+  return *this;
+}
+
+c_qualifierst &java_qualifierst::operator=(const c_qualifierst &c_other)
+{
+  auto other = dynamic_cast<const java_qualifierst *>(&c_other);
+  if(other != nullptr)
+    *this = *other;
+  else
+  {
+    annotations.clear();
+    c_qualifierst::operator=(c_other);
+  }
+  return *this;
+}
+
+std::unique_ptr<c_qualifierst> java_qualifierst::clone() const
+{
+  auto other = util_make_unique<java_qualifierst>(ns);
+  *other = *this;
+  return std::move(other);
+}
+
+const std::vector<java_annotationt> &java_qualifierst::get_annotations() const
+{
+  return annotations;
+}
+
+std::size_t java_qualifierst::count() const
+{
+  return c_qualifierst::count() + annotations.size();
+}
+
+void java_qualifierst::clear()
+{
+  c_qualifierst::clear();
+  annotations.clear();
+}
+
+void java_qualifierst::read(const typet &src)
+{
+  c_qualifierst::read(src);
+  auto &annotated_type = static_cast<const annotated_typet &>(src);
```

Comment:
There should be a `to_annotated_type`.

Reply:
Even if there's no possible run-time check for this?

Reply:
Yes, because it puts all the type casts in one place where maybe, some day, the type hierarchy can be changed. Or additional checks can be added.